tests/components/zwave_js/test_init.py

```diff
@@ -738,63 +842,279 @@ async def test_node_removed(hass, multisensor_6_state, client, integration):
+    assert device == dev_reg.async_get_device(identifiers={(DOMAIN, hank_device_id)})
+    assert not dev_reg.async_get_device(identifiers={(DOMAIN, multisensor_6_device_id)})
     assert device.manufacturer == "HANK Electronics Ltd."
+    assert device.model == "HKZW-SO01"
+
+    assert not hass.states.get(AIR_TEMPERATURE_SENSOR)
+    assert hass.states.get("switch.smart_plug_with_two_usb_ports")
+
+
+async def test_node_model_change(hass, zp3111, zp3111_state, client, integration):
+    """Test that when a node's model is changed due to an updated device config file, the device and entities are not removed."""
+    # This is not 100% realistic test, since the model change would be seen when the integration is loaded, not via a runtime event. The same registration code path is used though, so it practically similar
+    dev_reg = dr.async_get(hass)
+    er_reg = er.async_get(hass)
+    zp3111_state = deepcopy(zp3111_state)
+
+    device_id = f"{client.driver.controller.home_id}-{zp3111.node_id}"
+    device_id_ext = f"{device_id}-{zp3111.manufacturer_id}:{zp3111.product_type}:{zp3111.product_id}"
+
+    device = dev_reg.async_get_device(identifiers={(DOMAIN, device_id)})
+    assert device
+    assert device == dev_reg.async_get_device(identifiers={(DOMAIN, device_id_ext)})
+    dev_id = device.id
+
+    motion_entity = "binary_sensor.4_in_1_sensor_home_security_motion_detection"
+    state = hass.states.get(motion_entity)
+    assert state
+    assert state.name == "4-in-1 Sensor: Home Security - Motion detection"
+
+    dev_reg.async_update_device(device.id, name_by_user="Custom Device Name")
+    device = dev_reg.async_get_device(identifiers={(DOMAIN, device_id)})
+    assert device
+    assert device.id == dev_id
+    assert device == dev_reg.async_get_device(identifiers={(DOMAIN, device_id_ext)})
+    assert device.name == "4-in-1 Sensor"
+    assert device.name_by_user == "Custom Device Name"
+
+    custom_entity = "binary_sensor.custom_motion_sensor"
+    er_reg.async_update_entity(
+        motion_entity, new_entity_id=custom_entity, name="Custom Entity Name"
+    )
+    await hass.async_block_till_done()
+    assert not hass.states.get(motion_entity)
+    state = hass.states.get(custom_entity)
+    assert state
+    assert state.name == "Custom Entity Name"
+
+    # simulate device config file changes
+    zp3111_state["deviceConfig"]["label"] = "New Device Model"
+    zp3111_state["deviceConfig"]["description"] = "New Device Description"
+
+    # Trigger a node add event which will call async_on_node_ready. We can't use
+    # a node ready event directly because it's only listened to once. Neither of
+    # these events would occur in real life in this sequence.
+    event = Event(
+        type="node added",
+        data={
+            "source": "controller",
+            "event": "node added",
+            "node": zp3111_state,
+        },
+    )
+    client.driver.receive_event(event)
+    await hass.async_block_till_done()
```

Comment:
Emitting this event seems to call `async_on_node_added` two times, and I'm not sure why.

Reply:
It doesn't seem to run twice for me when running the test.

I get an error for non unique id though:
```
DEBUG    homeassistant.core:core.py:721 Bus:Handling <Event device_registry_updated[L]: action=update, device_id=e91f0d0915ef550241699fbdc78a683e>
DEBUG    homeassistant.core:core.py:721 Bus:Handling <Event entity_registry_updated[L]: action=update, entity_id=binary_sensor.custom_motion_sensor, changes=name=None, entity_id=binary_sensor.4_in_1_sensor_home_security_motion_detection, old_entity_id=binary_sensor.4_in_1_sensor_home_security_motion_detection>
DEBUG    homeassistant.core:core.py:721 Bus:Handling <Event state_changed[L]: entity_id=binary_sensor.4_in_1_sensor_home_security_motion_detection, old_state=<state binary_sensor.4_in_1_sensor_home_security_motion_detection=off; device_class=motion, friendly_name=4-in-1 Sensor: Home Security - Motion detection @ 2021-12-01T15:01:09.684376+00:00>, new_state=None>
DEBUG    homeassistant.core:core.py:721 Bus:Handling <Event state_changed[L]: entity_id=binary_sensor.custom_motion_sensor, old_state=None, new_state=<state binary_sensor.custom_motion_sensor=off; device_class=motion, friendly_name=Custom Entity Name @ 2021-12-01T15:01:09.688498+00:00>>
DEBUG    homeassistant.components.zwave_js:__init__.py:238 Processing node Node(node_id=22)
DEBUG    homeassistant.core:core.py:721 Bus:Handling <Event device_registry_updated[L]: action=update, device_id=e91f0d0915ef550241699fbdc78a683e>
DEBUG    homeassistant.core:core.py:721 Bus:Handling <Event entity_registry_updated[L]: action=update, entity_id=sensor.4_in_1_sensor_node_status, changes=original_name=4-in-1 Sensor: Node Status>
ERROR    homeassistant.components.sensor:entity_platform.py:586 Platform zwave_js does not generate unique IDs. ID 3245146787.22.node_status already exists - ignoring sensor.4_in_1_sensor_node_status
```

Haven't looked into detail why yet.

Reply:
I think that's a result of the tests that are sending `'node added'` events for devices that already exist.

Reply:
Can we make the test more realistic? Or do you mean that this would happen for real too?

Reply:
I think it will happen normally, but I'll need to review the test cases to be sure. Once the device is originally created along with the node status sensor, any later `node added` event will try to create the status entity. 

I think this test could be more realistic, but I'm not sure the right way to do it. There's maybe two ways:

1. load the integration and nodes, then simulate a disconnect event (?), followed by the initial driver node state message with the changed node info.
2. Populate the device registry before the integration is loaded, and the test uses the "non-ready" or changed node info instead

I think either of those would require a different mock client fixture?

Reply:
If the `node added` event is emitted when replacing failed nodes, the test is realistic as it is, right?

Do we need to re-register device and entity listeners on the replaced node when it's added?

For all other entities than the node status entity we have a dict of sets that registers already discovered node values to avoid discovering them again. I guess this is why only the node status entity complains about non unique id.

Reply:
> If the node added event is emitted when replacing failed nodes, the test is realistic as it is, right?

This particular test is trying to simulate the case where a device configuration has changed the model name, it's not for failed nodes. For example, in release X, a device model is "Foo", and in release Y it's renamed to "Bar". When you upgrade zwave-js, HA will throw away the device because of the model mismatch. The only time this would be seen is when the integration is loaded. In either of those cases, you're getting the initial node state dump from the connection. It's not the event that is adding the nodes, it's `start_platforms()` that is invoking `async_on_node_added()` for all of the nodes. I'm using the "node added" event to trigger a call to `async_on_node_added()`, since I can't directly access the function.

A more realistic test would be a fixture that populates the device registry before the integration is loaded, and the initial node state is seeded with a different model. I'm not quite sure how to accomplish that, is it possible?

In regards to replacing a failed node with the same device, I don't think the new node will ever have a different model from the old one. The device config files are loaded when zwave-js starts up and aren't changed while it's online, so replacing a node would use the same info as the old one.

> Do we need to re-register device and entity listeners on the replaced node when it's added?

I'm not sure. Would that be necessary if the underlying Node instance in the library were replaced? Is there something I could add to the test to confirm? I can do some more testing of that situation with a real replacement.

> For all other entities than the node status entity we have a dict of sets that registers already discovered node values to avoid discovering them again. I guess this is why only the node status entity complains about non unique id.

Makes sense. Is it something we can just ignore for now then?

Reply:
> A more realistic test would be a fixture that populates the device registry before the integration is loaded, and the initial node state is seeded with a different model. I'm not quite sure how to accomplish that, is it possible?

Yes. We can create entries in the device registry manually in the test or in a fixture, then set up the integration with `hass.config_entries.async_setup` instead of using the `integration` fixture. The node state can be seeded with whatever we want by setting the appropriate node in the driver list on the client before setting up the integration.

> Would that be necessary if the underlying Node instance in the library were replaced? Is there something I could add to the test to confirm?

I looked in the library and I think the event listeners will still work since they are based on the event string names and not node or value instances. The problem is in the existing entities that have saved a reference to the client node value instance. I don't see that we're reusing the old values when replacing a node if it's the "node added" and "node ready" events that are used also for replacing a node. This means that all the entities of the replaced node will have stale value references.

We should probably register a listener in the entity that removes the entity instance including the discovered value ids, when the node is replaced. Then the node ready event should discover the value and entity again when it comes in and the entity will have the correct value references.

We currently listen for `"node removed"` in `__init__.py` and remove the device if it was not replaced. That will also remove the entities and the entity entries from the registry. So we should listen for that event somewhere, in that place or in the entities, but handle the replace case, and only remove the entity instances in that case.

Reply:
I've attempted to try and reload the integration in the model change test, so it's configured as expected by the fixtures, and I just modify the node values. Loading the integration again seems to do what's expected. Does that seem reasonable?

Haven't looked at the event listener stuff yet.

Reply:
Yes, looks good! 👍